File: app/src/main/java/com/nesib/countriesapp/database/DatabaseHelper.java
package com.nesib.countriesapp.database;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;

import com.nesib.countriesapp.Constants;
import com.nesib.countriesapp.models.Country;
import com.nesib.countriesapp.models.Currency;
import com.nesib.countriesapp.models.Language;

import java.util.ArrayList;

public class DatabaseHelper extends SQLiteOpenHelper {
    private static DatabaseHelper instance;

    public static synchronized DatabaseHelper getInstance(Context context) {
        if (instance == null) {
            instance = new DatabaseHelper(context);
        }
        return instance;
    }

    public DatabaseHelper(Context context) {
        super(context, Constants.DB_NAME, null, 1);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        // CREATE TABLE countries_table(id,flag,alpha_code,full_name,region,area,population)
        db.execSQL("CREATE TABLE " +
                Constants.TABLE_NAME + "(" +
                "id INTEGER PRIMARY KEY," +
                Constants.COL_FLAG + " TEXT," +
                Constants.COL_ALPHA3_CODE + " TEXT," +
                Constants.COL_FULL_NAME + " TEXT," +
                Constants.COL_NAME + " TEXT," +
                Constants.COL_REGION + " TEXT," +
                Constants.COL_POPULATION + " NUMBER," +
                Constants.COL_AREA + " NUMBER," +
                Constants.COL_BORDERS + " TEXT," +
                Constants.COL_CAPITAL + " TEXT," +
                Constants.COL_LAT + " REAL," +
                Constants.COL_LNG + " REAL," +
                Constants.COL_CALLING_CODE + " TEXT," +
                Constants.COL_LANGUAGES + " TEXT," +
                Constants.COL_CURRENCIES + " TEXT" + ")"
        );
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int i, int i1) {
        db.execSQL("DROP TABLE IF EXISTS " + Constants.TABLE_NAME);
        onCreate(db);
    }

    public long addCountry(Country country) {
        Log.d("mytag", "addCountry: " + country);
        SQLiteDatabase db = getWritableDatabase();
        ContentValues values = new ContentValues();
        values.put(Constants.COL_FLAG, country.getFlag());
        values.put(Constants.COL_ALPHA3_CODE, country.getAlpha3Code());
        if (country.getAltSpellings().length > 1) {
            values.put(Constants.COL_FULL_NAME, country.getAltSpellings()[1]);
        } else {
            values.put(Constants.COL_FULL_NAME, country.getAltSpellings()[0]);
        }
        values.put(Constants.COL_NAME, country.getName());
        values.put(Constants.COL_REGION, country.getRegion());
        values.put(Constants.COL_POPULATION, country.getPopulation());
        values.put(Constants.COL_AREA, country.getArea());
        values.put(Constants.COL_CAPITAL, country.getCapital());

        values.put(Constants.COL_LAT, country.getLatlng()[0]);
        values.put(Constants.COL_LNG, country.getLatlng()[1]);
        values.put(Constants.COL_BORDERS, arrayToString(country.getBorders()));
        values.put(Constants.COL_CALLING_CODE, arrayToString(country.getCallingCodes()));

        Language[] languages = country.getLanguages();
        Currency[] currencies = country.getCurrencies();
        StringBuilder languageText = new StringBuilder();
        StringBuilder currencyText = new StringBuilder();
        for (Language lang : languages) {
            languageText.append(lang.getName()).append(",");
        }
        for (Currency cur : currencies) {
            currencyText.append(cur.getName()).append(",");
        }
        languageText.deleteCharAt(languageText.length() - 1);
        if (currencyText.length() > 0) {
            currencyText.deleteCharAt(currencyText.length() - 1);
        }

        values.put(Constants.COL_CURRENCIES, currencyText.toString());
        values.put(Constants.COL_LANGUAGES, languageText.toString());

        return db.insert(Constants.TABLE_NAME, null, values);
    }

    public String arrayToString(String[] array) {
        StringBuilder text = new StringBuilder();
        for (int i = 0; i < array.length; i++) {
            if (i != array.length - 1) {
                text.append(array[i]).append(",");
            } else {
                text.append(array[i]);
            }
        }
        return text.toString();
    }

    public void deleteCountry(Country country) {
        SQLiteDatabase db = getWritableDatabase();
        db.delete(Constants.TABLE_NAME, Constants.COL_FLAG + "=?", new String[]{country.getFlag()});
    }

    public ArrayList<Country> getCountryCodes() {
        ArrayList<Country> countryList = new ArrayList<>();
        int count = 0;
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT * FROM " + Constants.TABLE_NAME, null);
        if (cursor.moveToFirst()) {
            while (!cursor.isAfterLast()) {
                count++;
                Country country = new Country();
                country.setName(cursor.getString(cursor.getColumnIndex(Constants.COL_NAME)));
                country.setAlpha3Code(cursor.getString(cursor.getColumnIndex(Constants.COL_ALPHA3_CODE)));
                country.setFlag(cursor.getString(cursor.getColumnIndex(Constants.COL_FLAG)));
                country.setArea(cursor.getDouble(cursor.getColumnIndex(Constants.COL_AREA)));
                country.setPopulation(cursor.getLong(cursor.getColumnIndex(Constants.COL_POPULATION)));
                country.setRegion(cursor.getString(cursor.getColumnIndex(Constants.COL_REGION)));
                country.setAltSpellings(new String[]{cursor.getString(cursor.getColumnIndex(Constants.COL_FULL_NAME))});
                country.setCapital(cursor.getString(cursor.getColumnIndex(Constants.COL_CAPITAL)));
                // lat lng
                double lat = cursor.getDouble(cursor.getColumnIndex(Constants.COL_LAT));
                double lng = cursor.getDouble(cursor.getColumnIndex(Constants.COL_LNG));
                double latlng[] = new double[]{lat, lng};
                country.setLatlng(latlng);

                // Borders
                String borderText = cursor.getString(cursor.getColumnIndex(Constants.COL_BORDERS));
                String[] borders = new String[]{};
                if (borderText.length() > 0) {
                    borders = borderText.split(",");
                }
                country.setBorders(borders);

                // CallingCodes
                String callingCodesText = cursor.getString(cursor.getColumnIndex(Constants.COL_CALLING_CODE));
                String[] callingCodes = new String[]{};
                if (callingCodesText.length() > 0) {
                    callingCodes = callingCodesText.split(",");
                }
                country.setCallingCodes(callingCodes);

                // Currencies
                String currencyText = cursor.getString(cursor.getColumnIndex(Constants.COL_CURRENCIES));
                String[] currencyArray = new String[]{};
                if (currencyText.length() > 0) {
                    currencyArray = currencyText.split(",");
                }
                Currency[] currencies = new Currency[currencyArray.length];
                for (int i = 0; i < currencyArray.length; i++) {
                    Currency currency = new Currency(null, currencyArray[i], null);
                    currencies[i] = currency;
                }
                country.setCurrencies(currencies);


                // Languages
                String languageText = cursor.getString(cursor.getColumnIndex(Constants.COL_LANGUAGES));
                String[] languagesArray = new String[]{};
                if (languageText.length() > 0) {
                    languagesArray = languageText.split(",");
                }
                Language[] languages = new Language[languagesArray.length];
                for (int i = 0; i < languagesArray.length; i++) {
                    Language language = new Language(languagesArray[i]);
                    languages[i] = language;
                }
                country.setLanguages(languages);

                countryList.add(country);
                cursor.moveToNext();
            }
        }
        cursor.close();
        return countryList;
    }


}

File: app/src/main/java/com/nesib/countriesapp/database/DatabaseHelper.kt
package com.nesib.countriesapp.database

import android.database.sqlite.SQLiteOpenHelper
import android.database.sqlite.SQLiteDatabase
import com.nesib.countriesapp.models.Country
import android.content.ContentValues
import android.content.Context
import com.nesib.countriesapp.utils.Constants
import com.nesib.countriesapp.models.Currency
import com.nesib.countriesapp.models.Language
import java.lang.StringBuilder
import java.util.ArrayList
import kotlin.jvm.Synchronized

class DatabaseHelper(context: Context?) : SQLiteOpenHelper(context, Constants.DB_NAME, null, 1) {
    override fun onCreate(db: SQLiteDatabase) {
        // CREATE TABLE countries_table(id,flag,alpha_code,full_name,region,area,population)
        db.execSQL(
            "CREATE TABLE " +
                    Constants.TABLE_NAME + "(" +
                    "id INTEGER PRIMARY KEY," +
                    Constants.COL_FLAG + " TEXT," +
                    Constants.COL_ALPHA3_CODE + " TEXT," +
                    Constants.COL_FULL_NAME + " TEXT," +
                    Constants.COL_NAME + " TEXT," +
                    Constants.COL_REGION + " TEXT," +
                    Constants.COL_POPULATION + " NUMBER," +
                    Constants.COL_AREA + " NUMBER," +
                    Constants.COL_BORDERS + " TEXT," +
                    Constants.COL_CAPITAL + " TEXT," +
                    Constants.COL_LAT + " REAL," +
                    Constants.COL_LNG + " REAL," +
                    Constants.COL_CALLING_CODE + " TEXT," +
                    Constants.COL_LANGUAGES + " TEXT," +
                    Constants.COL_CURRENCIES + " TEXT" + ")"
        )
    }

    override fun onUpgrade(db: SQLiteDatabase, i: Int, i1: Int) {
        db.execSQL("DROP TABLE IF EXISTS " + Constants.TABLE_NAME)
        onCreate(db)
    }

    fun addCountry(country: Country) {
        val db = writableDatabase
        val values = ContentValues()
        values.put(Constants.COL_FLAG, country.flag)
        values.put(Constants.COL_ALPHA3_CODE, country.alpha3Code)
        if (country.altSpellings.size > 1) {
            values.put(Constants.COL_FULL_NAME, country.altSpellings[1])
        } else {
            values.put(Constants.COL_FULL_NAME, country.altSpellings[0])
        }
        values.put(Constants.COL_NAME, country.name)
        values.put(Constants.COL_REGION, country.region)
        values.put(Constants.COL_POPULATION, country.population)
        values.put(Constants.COL_AREA, country.area)
        values.put(Constants.COL_CAPITAL, country.capital)
        values.put(Constants.COL_LAT, country.latlng[0])
        values.put(Constants.COL_LNG, country.latlng[1])
        values.put(Constants.COL_BORDERS, arrayToString(country.borders))
        values.put(Constants.COL_CALLING_CODE, arrayToString(country.callingCodes))
        val languages = country.languages
        val currencies = country.currencies
        val languageText = StringBuilder()
        val currencyText = StringBuilder()
        for (lang in languages) {
            languageText.append(lang.name).append(",")
        }
        for (cur in currencies) {
            currencyText.append(cur.name).append(",")
        }
        languageText.deleteCharAt(languageText.length - 1)
        if (currencyText.isNotEmpty()) {
            currencyText.deleteCharAt(currencyText.length - 1)
        }
        values.put(Constants.COL_CURRENCIES, currencyText.toString())
        values.put(Constants.COL_LANGUAGES, languageText.toString())
        db.insert(Constants.TABLE_NAME, null, values)
    }

    private fun arrayToString(array: List<String>): String {
        val text = StringBuilder()
        for (i in array.indices) {
            if (i != array.size - 1) {
                text.append(array[i]).append(",")
            } else {
                text.append(array[i])
            }
        }
        return text.toString()
    }

    fun deleteCountry(country: Country) {
        val db = writableDatabase
        db.delete(Constants.TABLE_NAME, Constants.COL_FLAG + "=?", arrayOf(country.flag))
    }

    // lat lng
    val countryCodes:

    // Borders

    // CallingCodes

    // Currencies


    // Languages
            ArrayList<Country>
        get() {
            val countryList = ArrayList<Country>()
            val db = readableDatabase
            val cursor = db.rawQuery("SELECT * FROM " + Constants.TABLE_NAME, null)
            if (cursor.moveToFirst()) {
                while (!cursor.isAfterLast) {
                    val country = Country()
                    country.name = cursor.getString(cursor.getColumnIndexOrThrow(Constants.COL_NAME))
                    country.alpha3Code =
                        cursor.getString(cursor.getColumnIndexOrThrow(Constants.COL_ALPHA3_CODE))
                    country.flag = cursor.getString(cursor.getColumnIndexOrThrow(Constants.COL_FLAG))
                    country.area = cursor.getDouble(cursor.getColumnIndexOrThrow(Constants.COL_AREA))
                    country.population =
                        cursor.getLong(cursor.getColumnIndexOrThrow(Constants.COL_POPULATION))
                    country.region = cursor.getString(cursor.getColumnIndexOrThrow(Constants.COL_REGION))
                    country.altSpellings = listOf(
                        cursor.getString(
                            cursor.getColumnIndexOrThrow(
                                Constants.COL_FULL_NAME
                            )
                        )
                    )
                    country.capital = cursor.getString(cursor.getColumnIndexOrThrow(Constants.COL_CAPITAL))
                    // lat lng
                    val lat = cursor.getDouble(cursor.getColumnIndexOrThrow(Constants.COL_LAT))
                    val lng = cursor.getDouble(cursor.getColumnIndexOrThrow(Constants.COL_LNG))
                    val latlng = listOf<Double>(lat,lng)
                    country.latlng = latlng

                    // Borders
                    val borderText = cursor.getString(cursor.getColumnIndexOrThrow(Constants.COL_BORDERS))
                    var borders = listOf<String>()
                    if (borderText.isNotEmpty()) {
                        borders = borderText.split(",".toRegex())
                    }
                    country.borders = borders

                    // CallingCodes
                    val callingCodesText =
                        cursor.getString(cursor.getColumnIndexOrThrow(Constants.COL_CALLING_CODE))
                    var callingCodes = listOf<String>()
                    if (callingCodesText.isNotEmpty()) {
                        callingCodes = callingCodesText.split(",".toRegex())
                    }
                    country.callingCodes = callingCodes

                    // Currencies
                    val currencyText =
                        cursor.getString(cursor.getColumnIndexOrThrow(Constants.COL_CURRENCIES))
                    var currencyArray = arrayOf<String?>()
                    if (currencyText.isNotEmpty()) {
                        currencyArray = currencyText.split(",".toRegex()).toTypedArray()
                    }
                    val currencies = mutableListOf<Currency>()
                    for (i in currencyArray.indices) {
                        val currency = Currency(null, currencyArray[i], null)
                        currencies.add(currency)
                    }
                    country.currencies = currencies


                    // Languages
                    val languageText =
                        cursor.getString(cursor.getColumnIndexOrThrow(Constants.COL_LANGUAGES))
                    var languagesArray = arrayOf<String?>()
                    if (languageText.isNotEmpty()) {
                        languagesArray = languageText.split(",".toRegex()).toTypedArray()
                    }
                    val languages = mutableListOf<Language>()
                    for (i in languagesArray.indices) {
                        val language = Language(languagesArray[i]!!)
                        languages.add(language)
                    }
                    country.languages = languages
                    countryList.add(country)
                    cursor.moveToNext()
                }
            }
            cursor.close()
            return countryList
        }

    companion object {
        private var instance: DatabaseHelper? = null
        @JvmStatic
        @Synchronized
        fun getInstance(context: Context?): DatabaseHelper {
            if (instance == null) {
                instance = DatabaseHelper(context)
            }
            return instance!!
        }
    }
}